[Bug]: gsd --mode mcp fails with Cannot find module for @modelcontextprotocol/sdk/dist/cjs/server/stdio on SDK ≥   1.28.x

### Bug type

Regression (worked before, now fails)

### Summary

Error:
  [gsd] uncaught extension error (non-fatal): Cannot find module
    '/…/gsd-pi/node_modules/@modelcontextprotocol/sdk/dist/cjs/server/stdio'
  Error: Cannot find module '…/dist/cjs/server/stdio'
      at createEsmNotFoundErr (node:internal/modules/cjs/loader:1493:15)
      …
      at startMcpServer (…/gsd-pi/dist/mcp-server.js:26:44)

### Steps to reproduce

```
npm install -g gsd-pi
gsd --mode mcp
```

### Expected behavior

MCP server should run.

### Actual behavior

GSD shows the error message posted above.

### GSD version

2.67

### Node.js version

v24.14.0

### Operating system

Kubuntu 24

### Install method

npm global install

### Affected area

MCP / tools / extensions

### Model and provider

_No response_

### Logs, screenshots, and evidence

```shell

```

### Impact

_No response_

### Additional information


  Symptom

  Running gsd --mode mcp (or having Claude Code load the gsd MCP server from settings.json) fails with:

  [gsd] uncaught extension error (non-fatal): Cannot find module
    '.../gsd-pi/node_modules/@modelcontextprotocol/sdk/dist/cjs/server/stdio'

  followed by a similar error for dist/cjs/types.

  Root cause

  mcp-server.js uses createRequire + require.resolve() to resolve two SDK subpaths that are not explicitly listed in
   @modelcontextprotocol/sdk's package.json exports map:

  _require.resolve(`${MCP_PKG}/server/stdio`)  // → not in exports
  _require.resolve(`${MCP_PKG}/types`)          // → not in exports

  The SDK's exports map has a wildcard entry:
  "./*": { "require": "./dist/cjs/*" }

  On Node ≤22 this worked because the CJS resolver auto-appended .js. On Node v24, wildcard exports resolved through
   finalizeEsmResolution no longer auto-append .js, so dist/cjs/server/stdio (no extension) is searched and not
  found even though dist/cjs/server/stdio.js exists.

  Workaround

  Add explicit exports to node_modules/gsd-pi/node_modules/@modelcontextprotocol/sdk/package.json:

  "./server/stdio": {
      "import": "./dist/esm/server/stdio.js",
      "require": "./dist/cjs/server/stdio.js"
  },
  "./types": {
      "import": "./dist/esm/types.js",
      "require": "./dist/cjs/types.js"
  }

  Suggested fix

  Either:

  1. In gsd-pi mcp-server.js: Avoid require.resolve() for unlisted subpaths. Since StdioServerTransport isn't
  re-exported from @modelcontextprotocol/sdk/server, one option is to resolve the file path directly:
  const sdkRoot = path.dirname(_require.resolve(`${MCP_PKG}/package.json`));
  const stdioMod = await import(path.join(sdkRoot, 'dist/cjs/server/stdio.js'));
  2. Upstream in @modelcontextprotocol/sdk: Add explicit named exports for ./server/stdio and ./types so wildcard
  resolution isn't needed for these common subpaths.

--- a/src/mcp-server.ts
+++ b/src/mcp-server.ts
@@ -19,10 +19,10 @@ export interface McpToolDef {
 // MCP SDK subpath imports use wildcard exports (./*) that NodeNext resolves
 // at runtime but TypeScript cannot statically type-check. We construct the
 // specifiers dynamically so tsc treats them as `any`.
-// Use createRequire to resolve wildcard subpaths — CJS resolver auto-appends
-// .js, which the ESM wildcard export map does not (#3603).
-import { createRequire } from 'node:module'
-const _require = createRequire(import.meta.url)
+//
+// Use explicit .js subpaths for modules that are loaded dynamically at runtime.
+// Recent Node / SDK combinations do not reliably resolve the extensionless
+// wildcard targets for `server/stdio` and `types` (#3914).
 const MCP_PKG = '@modelcontextprotocol/sdk'
 
 /**
@@ -46,8 +46,8 @@ export async function startMcpServer(options: {
   const { tools, version = '0.0.0' } = options
 
   const serverMod = await import(`${MCP_PKG}/server`)
-  const stdioMod = await import(_require.resolve(`${MCP_PKG}/server/stdio`))
-  const typesMod = await import(_require.resolve(`${MCP_PKG}/types`))
+  const stdioMod = await import(`${MCP_PKG}/server/stdio.js`)
+  const typesMod = await import(`${MCP_PKG}/types.js`)
 
   const Server = serverMod.Server
   const StdioServerTransport = stdioMod.StdioServerTransport

--- a/src/tests/mcp-createRequire.test.ts
+++ b/src/tests/mcp-createRequire.test.ts
@@ -1,12 +1,9 @@
 /**
- * Regression test for #3603 — MCP server subpath imports via createRequire
+ * Regression test for #3914 — MCP server uses explicit .js SDK subpaths.
  *
- * The ESM wildcard export map in @modelcontextprotocol/sdk does not resolve
- * subpath imports correctly. The fix uses createRequire from node:module to
- * resolve wildcard subpaths via the CJS resolver which auto-appends .js.
- *
- * Structural verification test — reads source to confirm createRequire import
- * and _require.resolve usage exist.
+ * Extensionless wildcard exports for `server/stdio` and `types` do not resolve
+ * reliably across current Node / SDK combinations. The runtime import strings
+ * must include `.js`.
  */
 
 import { describe, test } from 'node:test';
@@ -20,29 +17,19 @@ const __dirname = dirname(__filename);
 
 const source = readFileSync(join(__dirname, '..', 'mcp-server.ts'), 'utf-8');
 
-describe('MCP server createRequire subpath resolution (#3603)', () => {
-  test('createRequire is imported from node:module', () => {
-    assert.match(source, /import\s*\{\s*createRequire\s*\}\s*from\s*['"]node:module['"]/,
-      'createRequire should be imported from node:module');
+describe('MCP server SDK subpath imports (#3914)', () => {
+  test('server/stdio import uses explicit .js subpath', () => {
+    assert.match(source, /await import\(`\$\{MCP_PKG\}\/server\/stdio\.js`\)/,
+      'server/stdio import should include the .js suffix');
   });
 
-  test('_require is created from import.meta.url', () => {
-    assert.match(source, /createRequire\(import\.meta\.url\)/,
-      '_require should be created using createRequire(import.meta.url)');
+  test('types import uses explicit .js subpath', () => {
+    assert.match(source, /await import\(`\$\{MCP_PKG\}\/types\.js`\)/,
+      'types import should include the .js suffix');
   });
 
-  test('_require.resolve is used for subpath imports', () => {
-    assert.match(source, /_require\.resolve\(/,
-      '_require.resolve should be used for subpath resolution');
-  });
-
-  test('server/stdio subpath uses _require.resolve', () => {
-    assert.match(source, /_require\.resolve\(`\$\{MCP_PKG\}\/server\/stdio`\)/,
-      'server/stdio import should use _require.resolve');
-  });
-
-  test('types subpath uses _require.resolve', () => {
-    assert.match(source, /_require\.resolve\(`\$\{MCP_PKG\}\/types`\)/,
-      'types import should use _require.resolve');
+  test('legacy createRequire-based resolution is gone', () => {
+    assert.doesNotMatch(source, /createRequire|_require\.resolve/,
+      'legacy createRequire-based subpath resolution should not remain');
   });
 });

--- a/src/tests/mcp-server.test.ts
+++ b/src/tests/mcp-server.test.ts
@@ -30,25 +30,11 @@ test('startMcpServer accepts the correct argument shape', async () => {
   assert.strictEqual(startMcpServer.length, 1, 'startMcpServer should accept one argument')
 })
 
-test('startMcpServer can be called with mock tools', async () => {
-  const { startMcpServer } = await import(distUrl('mcp-server.js'))
+test('compiled MCP runtime dependencies resolve with explicit .js subpaths', async () => {
+  const stdioMod = await import('@modelcontextprotocol/sdk/server/stdio.js')
+  const typesMod = await import('@modelcontextprotocol/sdk/types.js')
 
-  // Create a mock tool matching the McpToolDef interface
-  const mockTool = {
-    name: 'test_tool',
-    description: 'A test tool',
-    parameters: { type: 'object', properties: {} },
-    execute: async () => ({
-      content: [{ type: 'text', text: 'hello' }],
-    }),
-  }
-
-  // Verify the function can be called with the correct signature
-  // without throwing during argument validation. It will attempt to
-  // connect to stdin/stdout as an MCP transport, which won't work in
-  // a test environment, but the Server instance is created successfully.
-  assert.doesNotThrow(() => {
-    void startMcpServer({ tools: [mockTool], version: '0.0.0-test' })
-      .catch(() => { /* expected: no MCP client on stdin */ })
-  })
+  assert.strictEqual(typeof stdioMod.StdioServerTransport, 'function')
+  assert.ok(typesMod.ListToolsRequestSchema, 'ListToolsRequestSchema should be exported')
+  assert.ok(typesMod.CallToolRequestSchema, 'CallToolRequestSchema should be exported')
 })

